### ISSUE TITLE
Unable to use function return value as a type

### ISSUE THREAD

**Issue by patrick91** [microsoft/pyright#3564]
This was working in an earlier version of pyright, but it might still be something wrong with my code.

I have the following code that creates a python type with some metadata:

```python
def union(
    name: str, types: Tuple[Types, ...], *
) -> Union[Types]:
    # implementation is not important
    return None # type: ignore
```

when using this, say like so:

```python
UserOrError = union("UserOrError", (User, Error))

reveal_type(UserOrError)

x: UserOrError = User(name="Patrick")
```

I get the following error:

> Illegal type annotation: variable not allowed unless it is a type aliasPylance[reportGeneralTypeIssues](https://github.com/microsoft/pylance-release/blob/main/DIAGNOSTIC_SEVERITY_RULES.md#diagnostic-severity-rules)

while the reveal type works fine. Is there anything I can do to make this work? 

I tried with:

```python
UserOrError: TypeAlias = union("UserOrError", (User, Error))
```
and it works, but I'd love to be able to skip that step if possible

full code here:

```python
from typing import Union, Tuple, TypeVar, Type

class User:
    name: str


class Error:
    message: str

Types = TypeVar("Types", bound=Type)

def union(
    name: str, types: Tuple[Types, ...], *
) -> Union[Types]:
    # implementation is not important
    return None # type: ignore

UserOrError = union("UserOrError", (User, Error))

reveal_type(UserOrError)

x: UserOrError = User(name="Patrick")

reveal_type(x)
```

screenshot of the error: 
<img width="1008" alt="image" src="https://user-images.githubusercontent.com/667029/172944890-75c6473a-7f72-4d24-b169-d6bf003144e2.png">

**Comment by erictraut** [comment]
This isn't valid in the Python type system. `Union` is a special form that requires you to provide two or more subtypes. Consider that in Python 3.10 and newer, `Union` is replaced with a `|` operator (e.g. `str | int`).

Furthermore, type expressions (those used within type annotations, type alias definitions, etc.) cannot use arbitrary expressions such as function calls. There are certain special forms of function calls that are allowed, such as `NewType` and `TypeVar`, but arbitrary user-defined functions like `union` in your example are not allowed. So `UnionOrError` in your example is not considered a type alias; it's a regular variable. And regular variables cannot be used in type annotations.

Here is the proper way to define a type alias:

```python
UserOrError = Union[User, Error]
```
or
```python
UserOrError = User | Error
```
or
```python
UserOrError: TypeAlias = User | Error
```

**Comment by patrick91** [comment]
@erictraut thanks for explaining 😊
